### PR TITLE
Cakephp 3.x upgrade

### PR DESCRIPTION
**Pull request by notpron**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,67 +9,296 @@
...
+- Should a Workerprocess quit when there are no more tasks for it to execute (true = exit, false = keep running):
+
+    ```php
+    $config['Queue']['exitWhenNothingToDo'] = false;
+    ```
+
+- Minimum number of seconds before a cleanup run will remove a completed task (set to 0 to disable):
+
+    ```php
+    $config['Queue']['cleanupTimeout'] = 2592000; // 30 days
```
Comment (tersmitten) [inline README.md]: Do all the options match and still have the same defaults?

File: README.md [hunk 2]
```diff
@@ -9,67 +9,296 @@
 
 ## Requirements
 
-* CakePHP 2.9.0 or greater.
+* CakePHP 3.7.0 or greater.
 * PHP 7.0.0 or greater.
```
Comment (tersmitten) [inline README.md]: `7.1.0`

File: README.md [hunk 3]
```diff
@@ -9,67 +9,296 @@
...
+
+For sending emails, for example:
+
+```php
+// In your controller
+$this->loadModel('Queue.QueuedTasks');
+$this->QueuedTasks->createJob('Email', ['to' => 'user@example.org', ...]);
+
+// Somewhere in the model or lib
+TableRegistry::get('Queue.QueuedTasks')->createJob('Email',
```
Comment (tersmitten) [inline README.md]: Is this section (still) relevant?

File: phpstan.neon [hunk 4]
```diff
@@ -0,0 +1,6 @@
+parameters:
+    autoload_files:
+        - tests/bootstrap.php
+
+    ignoreErrors:
+        - '#Result of method .+QueueTaskInterface::run\(\) \(void\) is used.#'
```
Comment (tersmitten) [inline phpstan.neon]: Mssing `EOL`

File: tests/bootstrap.php [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    'driver' => getenv('db_class'),
+    'dsn' => getenv('db_dsn'),
+    'database' => getenv('db_database'),
+    'username' => getenv('db_username'),
+    'password' => getenv('db_password'),
+    'timezone' => 'UTC',
+    'quoteIdentifiers' => true,
+    'cacheMetadata' => true,
+
+]);
```
Comment (tersmitten) [inline tests/bootstrap.php]: Does running tests in an actual `app` work?

**Comment by tersmitten** [comment]
Do we want to keep the incorrect `po(t)` files?

**Comment by tersmitten** [comment]
The minimal PHP version is 7.1, do we want a add (more) typehints?

**Comment by tersmitten** [comment]
Most important: do we still have all the options and do they work as expected (and have the same defaults)?

File: composer.json [hunk 6]
```diff
@@ -14,13 +14,35 @@
...
+		"cakephp/cakephp": "^3.7"
+	},
+	"require-dev": {
+		"oefenweb/cakephp-codesniffer": "^3.0.0",
+		"phpmd/phpmd": "^2.6",
+		"phpstan/phpstan": "^0.10.3",
+		"phpunit/phpunit": "^5.7.14|^6.0",
+		"psy/psysh": "@stable",
+		"sebastian/phpcpd": "^3.0",
+		"dereuromark/cakephp-tools": "^1.9",
```
Comment (tersmitten) [inline composer.json]: Where do we use this for?

File: src/Shell/Task/AddInterface.php [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<?php
+namespace Queue\Shell\Task;
+
+/**
+ * Any task needs to at least implement add().
+ *
+ * @author Mark Scherer
+ * @license http://www.opensource.org/licenses/mit-license.php The MIT License
```
Comment (tersmitten) [inline src/Shell/Task/AddInterface.php]: Do we want to keep these kind of comments?

File: tests/bootstrap.php (hunk 5 above)
Comment (notpron) [reply tests/bootstrap.php]: It does

File: composer.json (hunk 6 above)
Comment (notpron) [reply composer.json]: Used by the unit tests to avoid creating wrappers

**Pull request merged**